AX: Collapse AccessibilityTableRow into AccessibilityRenderObject

hoffmanjoshua opened the pull request:
#### 77c558a9da15183e15277c9736703686df70151a
<pre>
AX: Collapse AccessibilityTableRow into AccessibilityRenderObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=297947">https://bugs.webkit.org/show_bug.cgi?id=297947</a>
<a href="https://rdar.apple.com/159245276">rdar://159245276</a>

Reviewed by Tyler Wilcock.

This moves all of the logic that used to live in AccessibilityTableRow into AccessibilityRenderObject
and AccessibilityNodeObject, as applicable. This takes advantage of the new rare data structure.

Notably, type checks to AccessibilityTableRow were replaced by a call to `isExposedTableRow`, formerly
`isTableRow`. This is equivalent to the old behavior, since that method was used in type check via the
type traits.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::selectedRows):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
(WebCore::AXObjectCache::handleChildrenChanged):
(WebCore::isAccessibilityTableRow): Deleted.
(WebCore::isAccessibilityARIAGridRow): Deleted.
* Source/WebCore/accessibility/AXObjectRareData.h:
(WebCore::AXObjectRareData::rowIndex const):
(WebCore::AXObjectRareData::setRowIndex):
* Source/WebCore/accessibility/AXTableHelpers.cpp:
(WebCore::AXTableHelpers::hasRowRole):
(WebCore::AXTableHelpers::isTableRowElement):
* Source/WebCore/accessibility/AXTableHelpers.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRole):
(WebCore::AccessibilityNodeObject::updateOwnedChildrenIfNecessary):
(WebCore::AccessibilityNodeObject::addChildren):
(WebCore::AccessibilityNodeObject::rowHeaders):
(WebCore::AccessibilityNodeObject::computeCellSlots):
(WebCore::AccessibilityNodeObject::parentTable const):
(WebCore::AccessibilityNodeObject::setRowIndex):
(WebCore::AccessibilityNodeObject::axColumnIndex const):
(WebCore::AccessibilityNodeObject::axRowIndex const):
(WebCore::AccessibilityNodeObject::axRowIndexText const):
(WebCore::AccessibilityNodeObject::disclosedRows):
(WebCore::AccessibilityNodeObject::disclosedByRow const):
(WebCore::AccessibilityNodeObject::isARIAGridRow const):
(WebCore::AccessibilityNodeObject::isARIATreeGridRow const):
(WebCore::AccessibilityNodeObject::isTableRow const):
(WebCore::AccessibilityNodeObject::parentTableIfExposedTableRow const):
(WebCore::AccessibilityNodeObject::updateOwnedChildren):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
(isType): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::isTableComponent):
(WebCore::AccessibilityObject::ignoredByRowAncestor const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::needsRareData const):
(WebCore::AccessibilityObject::isAccessibilityNodeObject const): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::computeIsIgnored const):
(WebCore::AccessibilityRenderObject::observableObject const):
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
(WebCore::AccessibilityRenderObject::addChildren):
* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
(WebCore::AccessibilityTableCell::parentRow const):
* Source/WebCore/accessibility/AccessibilityTableCell.h:
* Source/WebCore/accessibility/AccessibilityTableRow.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityTableRow.h: Removed.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::isAccessibilityNodeObject const):
(WebCore::AXIsolatedObject::isTableRow const):
(WebCore::AXIsolatedObject::parentTableIfExposedTableRow const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateDependentProperties):
(WebCore::convertToPropertyFlag):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeNames]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):

Canonical link: <a href="https://commits.webkit.org/299517@main">https://commits.webkit.org/299517@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee0db9f916c459874c4604938aff6be9d2b30275

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119248 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38931 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29585 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125479 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71313 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/781b587c-d66a-447a-b4ea-e0aecdf9bb21) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121126 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39627 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47512 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90599 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59956 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/1fa70801-a004-47aa-8f48-efd569cf3c1e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122201 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31602 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106904 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71014 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2011fd38-8f62-422e-b59c-acad4275e05f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30645 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25017 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69129 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101049 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25202 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128492 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46160 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34912 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99164 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46525 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103114 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98945 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25151 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44408 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22414 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42733 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46026 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51710 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45492 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48842 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47182 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2359,6 +2366,157 @@ void AccessibilityNodeObject::recomputeIsExposableIfNecessary()
+    // only have "row" elements, but if not, we still should handle it gracefully by finding the right table.
+    for (RefPtr parent = parentObject(); parent; parent = parent->parentObject()) {
+        if (parent->isTable()) {
+            bool isNonGridRowOrValidAriaTable = !isARIAGridRow() || parent->isAriaTable() || elementName() == ElementName::HTML_tr;
```
Is it better to check elementName() == HTML_tr or is<HTMLTableRowElement>?

hoffmanjoshua replied:
I don't think there would be a difference in either (behavior or perf) either way. I lean toward tag name, since we could eventually share this implementation between the isolated and live tree, and we can cache tag names on the isolated tree. What do you think?

minorninth commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2359,6 +2366,157 @@ void AccessibilityNodeObject::recomputeIsExposableIfNecessary()
+    // The parent table might not be the direct ancestor of the row unfortunately. ARIA states that role="grid" should
+    // only have "row" elements, but if not, we still should handle it gracefully by finding the right table.
+    for (RefPtr parent = parentObject(); parent; parent = parent->parentObject()) {
+        if (parent->isTable()) {
```
Optional: I feel like we could safely exit the for loop early if we get anything other than a row group, div, span, or other generic role. If we get any other role then there's definitely no parent table and we could exit the loop early and stop walking up.

hoffmanjoshua replied:
I agree with this, but I'd like to do this in a follow up so I can add appropriate test coverage

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1311,8 +1292,8 @@ void AXObjectCache::handleChildrenChanged(AccessibilityObject& object)
-    else if (auto* axRow = dynamicDowncast<AccessibilityTableRow>(object)) {
-        if (RefPtr parentTable = dynamicDowncast<AccessibilityNodeObject>(axRow->parentTable()))
+    else if (RefPtr renderObject = dynamicDowncast<AccessibilityRenderObject>(object); renderObject && renderObject->isExposedTableRow()) {
+        if (RefPtr parentTable = dynamicDowncast<AccessibilityNodeObject>(renderObject->parentTable()))
```
Here, we call `isExposedTableRow()`, which calls `parentTable()`, which does an ancestry traversal. Then we immediately call `parentTable()` again in the next line, repeating the work.

Here's one idea, maybe you'll have a better one though.

What if what is currently called `isExposableTable()` is renamed to `parentTableIfExposableRow()`, doing as described in the method name. Then isExposableTable() can be implemented as `isExposableTable() { return parentTableIfExposableRow()` for places that really only need the bool value.

twilco replied:
And probably an even better idea than that: we already do an ancestry traversal in this method — can we just add some logic into that rather than having this branch at all?

hoffmanjoshua replied:
Updated with a new method: parentTableIfExposedTableRow

twilco commented on Source/WebCore/accessibility/AXObjectRareData.h:
```diff
@@ -70,6 +75,10 @@ class AXObjectRareData : public CanMakeCheckedPtr<AXObjectRareData> {
 
+    // Begin table-row-related fields.
+    unsigned m_rowIndex;
+    // Edn table-row-related fields.
```
Is this optimally placed as per the diagnostics from `dump-class-layout`? Seems like it would introduce padding here with the preceding bool?

hoffmanjoshua replied:
This is fitting inside the padding generated after this bool (7 bytes of padding, total object size of 72 bytes). If we move this to the other padding available in this method, before the member variables, the object is still 72 bytes, so either spot will be the most efficient. I'll move it to the top for simplicity, and so we have all the padding in one place.

hoffmanjoshua replied:
Either here or at the beginning of the member vars would yield the same size: 72 bytes. There are 7 extra bytes here after these bool that this fits into the end of

twilco commented on Source/WebCore/accessibility/AXTableHelpers.cpp:
```diff
@@ -76,6 +76,24 @@ bool isTableRole(AccessibilityRole role)
+bool isARIAGridRow(Element& element)
+{
+    return hasRole(element, "row"_s);
+}
```
The method name would be a bit more descriptive at each callsite if it was `hasRowRole`, since that's all this function does.

twilco commented on Source/WebCore/accessibility/AXTableHelpers.cpp:
```diff
@@ -76,6 +76,24 @@ bool isTableRole(AccessibilityRole role)
+
+    bool isNonAnonymousRenderTable = is<RenderTableRow>(renderer.get()) && !isAnonymous;
+    bool ariaRoleIsValidForRow = hasRole(element, "row"_s) || hasRole(element, nullAtom());
+    return (isNonAnonymousRenderTable || is<HTMLTableRowElement>(element) || AXTableHelpers::isARIAGridRow(element)) && ariaRoleIsValidForRow;
```
I think this would both be more efficient and more clear if we just early-returned like this, since we require `ariaRoleIsValidForRow` for all the other conditions and thus shouldn't compute them first.

`isNonAnonymousRenderTable` is a bit of confusing name when we're checking for a RenderTableRow and not a RenderTable.

```cpp
if (hasRowRole(element))
    return true;

if (!hasRole(element, nullAtom()) {
    // This has a non-row role, so it shouldn't be considered a row.
    return false;
}

if (is<RenderTableRow>(renderer.get()) && !isAnonymous)
    return true;

return is<HTMLTableRowElement>(element));
```

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -696,10 +698,13 @@ void AccessibilityNodeObject::clearChildren()
 }
 
-void AccessibilityNodeObject::updateOwnedChildren()
+bool AccessibilityNodeObject::updateOwnedChildren()
```
Should we update the method name to hint at what the returned bool means?

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2202,15 +2209,15 @@ unsigned AccessibilityNodeObject::computeCellSlots()
                 processRowDescendingIfNeeded(child.get());
-        } else
-            processRow(dynamicDowncast<AccessibilityTableRow>(axObject));
+        } else if (axObject.isExposedTableRow())
```
Can we add a FIXME here indicating we do an ancestry traversal in `needsToDescend` via `!axObject.isExposedTableRow()`, then do it again here? Or even better, maybe we can just fix it. I feel like it's valid for these callsites to just use `AXTableHelpers::isTableRowElement(*element)`, or maybe we can have a method on `AccessibilityNodeObject` called is `isTableRow()` that we could call here. We wouldn't be computing the cellslots (this method) if the containing table wasn't exposed, seems wasteful to do an ancestry traversal to check it.

hoffmanjoshua replied:
Fixed this with a new method, `isTableRow()`

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2232,8 +2239,8 @@ unsigned AccessibilityNodeObject::computeCellSlots()
-                RefPtr tableRow = dynamicDowncast<AccessibilityTableRow>(cache->getOrCreate(row.get()));
-                processRow(tableRow.get());
+                if (RefPtr tableRow = cache->getOrCreate(row.get()); tableRow && tableRow->isExposedTableRow())
+                    processRow(dynamicDowncast<AccessibilityRenderObject>(tableRow).get());
```
Same comment here. Is checking `isExposedTableRow()` really necessary here, or `isTableRow()` enough?

hoffmanjoshua replied:
Yep, fixed!

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2359,6 +2366,157 @@ void AccessibilityNodeObject::recomputeIsExposableIfNecessary()
+{
+    RefPtr element = this->element();
+    if (!element && !AXTableHelpers::isTableRowElement(*element))
+        return nullptr;
```
This if-statement isn't correct. It only dereferences the element if it's null, guaranteeing a crash.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2359,6 +2366,157 @@ void AccessibilityNodeObject::recomputeIsExposableIfNecessary()
+
+    // The parent table might not be the direct ancestor of the row unfortunately. ARIA states that role="grid" should
+    // only have "row" elements, but if not, we still should handle it gracefully by finding the right table.
+    for (RefPtr parent = parentObject(); parent; parent = parent->parentObject()) {
```
`ancestor` would be a better name than `parent`, since `parent` is only an accurate name for the first iteration of the loop.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2359,6 +2366,157 @@ void AccessibilityNodeObject::recomputeIsExposableIfNecessary()
+    if (!element && !AXTableHelpers::isTableRowElement(*element))
+        return nullptr;
+
+    // The parent table might not be the direct ancestor of the row unfortunately. ARIA states that role="grid" should
```
"Parent table" (which is used in several places) is really not great terminology, considering situations like the comment is describing. I know this existed before your PR. Could you please add a FIXME to the `parentTable()` method suggesting a different name, like `containingTable`? And maybe the FIXME can mention the need to do a search for "parent table" comments when it gets addressed?

hoffmanjoshua replied:
FIXME added

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2359,6 +2366,157 @@ void AccessibilityNodeObject::recomputeIsExposableIfNecessary()
+    // have an aria-level of plus 1 from this row.
+    Ref parent = *parentObjectUnignored();
+    if (!parent->isExposableTable())
+        return disclosedRows;
```
I know you didn't add this code, but the unconditional deref seems unnecessary. Maybe we can fix it while we're here?

```cpp
RefPtr parent = parentObjectUnignored();
if (!parent || !parent->isExposableTable())
    return { };

AccessibilityChildrenVector disclosedRows;
```

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2359,6 +2366,157 @@ void AccessibilityNodeObject::recomputeIsExposableIfNecessary()
+    // that is aria-level subtract 1 from this row.
+    RefPtr parent = dynamicDowncast<AccessibilityNodeObject>(parentObjectUnignored());
+    if (!parent->isExposableTable())
+        return nullptr;
```
Missing null check here I think

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2359,6 +2366,157 @@ void AccessibilityNodeObject::recomputeIsExposableIfNecessary()
+        return nullptr;
+
+    for (int k = index - 1; k >= 0; --k) {
+        Ref row = allRows[k];
```
This adds unnecessary ref-count churn, maybe we can just use `allRows[k]` as needed? Or make this a `Ref&`?

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2359,6 +2366,157 @@ void AccessibilityNodeObject::recomputeIsExposableIfNecessary()
+
+bool AccessibilityNodeObject::isExposedTableRow() const
+{
+    // TODO: put this into a static method
```
Leftover TODO ☑️

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -4327,10 +4327,10 @@ bool AccessibilityObject::ignoredByRowAncestor() const
     });
 
-    return ancestor && ancestor->isTableRow();
+    return ancestor && ancestor->isExposedTableRow();
```
Another place where we duplicate ancestry traversals via `isExposedTableRow`. Maybe we can use a local bool variable to avoid duplicating that?

```cpp
bool AccessibilityObject::ignoredByRowAncestor() const
{
    bool wasExposedRow = false;
    RefPtr ancestor = Accessibility::findAncestor<AccessibilityObject>(*this, false, [&] (const AccessibilityObject& ancestor) {
        if (ancestor.isTableCell() || ancestor.isTable())
            return true;
        if (ancestor.isExposedTableRow())
            wasExposedRow = true;
        return wasExposedRow;
    });
    return ancestor && wasExposedRow;
}
```

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -2770,6 +2781,16 @@ void AccessibilityRenderObject::addChildren()
+            m_subtreeDirty = false;
+            // Skip the normal child building process, but still handle aria-colindex below
+        }
+    }
```
Do any tests fail if we remove this? This logic has always seemed wrong to me. Just because an object has aria-owns children doesn't mean it can't add DOM children. And this `bool usedOwnedObjects` adds a lot of complication to this method that's already pretty complicated, so would be nice to not have to do it

hoffmanjoshua replied:
Removed and all tests pass (with and without ITM), so I'll take this out.

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -2814,76 +2835,92 @@ void AccessibilityRenderObject::addChildren()
+    }
+
+    // Handle aria-colindex for all table rows (whether using owned objects or not).
+    if (isExposedTableRow()) {
```
We shouldn't call `isExposedTableRow()` twice in the span of one method call considering it does an ancestry traversal.

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -2814,76 +2835,92 @@ void AccessibilityRenderObject::addChildren()
+    // Handle aria-colindex for all table rows (whether using owned objects or not).
+    if (isExposedTableRow()) {
+        std::optional colIndex = axColumnIndex();
+        if (colIndex) {
```
Could combine this into one line:

```cpp
if (std::optional colIndex = axColumnIndex()) {
```

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -2814,6 +2825,7 @@ void AccessibilityRenderObject::addChildren()
         return false;
     };
 
+    // Don't do all of this extra work for objects like table rows if they use the aria-own objects.
```
I'm confused, what is this comment referring to exactly? Should it just be removed?

hoffmanjoshua replied:
Yep, this was leftover from when we would do the owned children update, it can be removed now.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -697,10 +699,13 @@ void AccessibilityNodeObject::clearChildren()
 }
 
-void AccessibilityNodeObject::updateOwnedChildren()
+bool AccessibilityNodeObject::updateOwnedChildrenIfNecessary()
```
Does anything use the bool this returns? If not, maybe we also shouldn't change the method name?

hoffmanjoshua replied:
Not anymore, fixed

twilco commented on Source/WebCore/accessibility/AXTableHelpers.cpp:
```diff
@@ -76,6 +76,33 @@ bool isTableRole(AccessibilityRole role)
+    CheckedPtr renderer = element.renderer();
+#if USE(ATSPI)
+    isAnonymous = renderer && renderer->isAnonymous();
+#endif
```
Super small nit, sorry I should've pointed it out last time I reviewed this part 😓 but maybe we can move these lines after the role checks, since they don't get used it we hit either of the first two branches?

hoffmanjoshua replied:
updated!

hoffmanjoshua committed:
AX: Collapse AccessibilityTableRow into AccessibilityRenderObject
https://bugs.webkit.org/show_bug.cgi?id=297947
rdar://159245276

Reviewed by Tyler Wilcock.

This moves all of the logic that used to live in AccessibilityTableRow into AccessibilityRenderObject
and AccessibilityNodeObject, as applicable. This takes advantage of the new rare data structure.

Notably, type checks to AccessibilityTableRow were replaced by a call to `isExposedTableRow`, formerly
`isTableRow`. This is equivalent to the old behavior, since that method was used in type check via the
type traits.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::selectedRows):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
(WebCore::AXObjectCache::handleChildrenChanged):
(WebCore::isAccessibilityTableRow): Deleted.
(WebCore::isAccessibilityARIAGridRow): Deleted.
* Source/WebCore/accessibility/AXObjectRareData.h:
(WebCore::AXObjectRareData::rowIndex const):
(WebCore::AXObjectRareData::setRowIndex):
* Source/WebCore/accessibility/AXTableHelpers.cpp:
(WebCore::AXTableHelpers::hasRowRole):
(WebCore::AXTableHelpers::isTableRowElement):
* Source/WebCore/accessibility/AXTableHelpers.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRole):
(WebCore::AccessibilityNodeObject::updateOwnedChildrenIfNecessary):
(WebCore::AccessibilityNodeObject::addChildren):
(WebCore::AccessibilityNodeObject::rowHeaders):
(WebCore::AccessibilityNodeObject::computeCellSlots):
(WebCore::AccessibilityNodeObject::parentTable const):
(WebCore::AccessibilityNodeObject::setRowIndex):
(WebCore::AccessibilityNodeObject::axColumnIndex const):
(WebCore::AccessibilityNodeObject::axRowIndex const):
(WebCore::AccessibilityNodeObject::axRowIndexText const):
(WebCore::AccessibilityNodeObject::disclosedRows):
(WebCore::AccessibilityNodeObject::disclosedByRow const):
(WebCore::AccessibilityNodeObject::isARIAGridRow const):
(WebCore::AccessibilityNodeObject::isARIATreeGridRow const):
(WebCore::AccessibilityNodeObject::isTableRow const):
(WebCore::AccessibilityNodeObject::parentTableIfExposedTableRow const):
(WebCore::AccessibilityNodeObject::updateOwnedChildren):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
(isType): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::isTableComponent):
(WebCore::AccessibilityObject::ignoredByRowAncestor const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::needsRareData const):
(WebCore::AccessibilityObject::isAccessibilityNodeObject const): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::computeIsIgnored const):
(WebCore::AccessibilityRenderObject::observableObject const):
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
(WebCore::AccessibilityRenderObject::addChildren):
* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
(WebCore::AccessibilityTableCell::parentRow const):
* Source/WebCore/accessibility/AccessibilityTableCell.h:
* Source/WebCore/accessibility/AccessibilityTableRow.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityTableRow.h: Removed.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::isAccessibilityNodeObject const):
(WebCore::AXIsolatedObject::isTableRow const):
(WebCore::AXIsolatedObject::parentTableIfExposedTableRow const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateDependentProperties):
(WebCore::convertToPropertyFlag):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeNames]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):

Canonical link: https://commits.webkit.org/299517@main
